ValveResourceFormat/Resource/ResourceTypes/KV3.cs

```diff
@@ -0,0 +1,592 @@
+            {
+                stateStack.Pop();
+            }
+
+            currentString.Append(c);
+        }
+    }
+
+    //Data structure to contain the file contents
+    public class KVFile
+    {
+        public String Header;
+        public KVObject Root;
+
+        public KVFile(String header, KVObject root)
+        {
+            this.Header = header;
+            this.Root = root;
+        }
+
+        //Serialize the contents of the data structure
+        public String Serialize()
+        {
+            StringBuilder stringBuilder = new StringBuilder();
+            stringBuilder.AppendLine(Header);
+            Root.Serialize(stringBuilder, 1);
+            return stringBuilder.ToString();
+        }
+    }
+
+    //Datastructure for a KV Object
+    public class KVObject
+    {
+        public String key;
+        public Dictionary<String, KVValue> Properties;
+
+        public KVObject(String name)
+        {
+            key = name;
+            Properties = new Dictionary<String, KVValue>();
+        }
+
+        //Add a property to the structure
+        public virtual void AddProperty(String name, KVValue value)
+        {
+            Properties.Add(name, value);
+        }
+
+        //Serialize the contents of the KV object
+        public virtual void Serialize(StringBuilder stringBuilder, int indent)
+        {
+            stringBuilder.Append("{\n");
+
+            foreach (var pair in Properties)
+            {
+                for (var i = 0; i < indent; i++)
+                {
+                    stringBuilder.Append("\t");
+                }
+                
+                stringBuilder.Append(pair.Key);
+                stringBuilder.Append(" = ");
+
+                PrintValue(stringBuilder, pair.Value.Type, pair.Value.Value, indent);
+                
+                stringBuilder.Append(Environment.NewLine);
+            }
+
+            for (var i = 0; i < indent-1; i++)
+            {
+                stringBuilder.Append("\t");
+            }
+            stringBuilder.Append("}");
+        }
+
+        //Print a value in the correct representation
+        protected void PrintValue(StringBuilder stringBuilder, Type type, Object value, int indent) {
+            if (type == typeof(KVObject))
+            {
+                ((KVObject)value).Serialize(stringBuilder, indent + 1);
+            }
+            else if (type == typeof(FlaggedString))
+            {
+                stringBuilder.Append((String)value);
+            }
+            else if (type == typeof(String))
+            {
+                String str = (String)value;
+
+                if (str.Contains("\n"))
+                {
+                    stringBuilder.Append("\"\"\"");
+                    stringBuilder.Append(str);
+                    stringBuilder.Append("\"\"\"");
+                }
+                else
+                {
+                    stringBuilder.Append("\"");
+                    stringBuilder.Append(str);
```

Comment:
Not if you assume the input is a KV file too since it should already escaped, you are right in the fact that if you put in a custom string this should probably be done, so depends on the usecase.